File: demos/langchain-tool-agent/src/tools/calculator.ts
import { tool } from '@langchain/core/tools'
import * as z from 'zod'

export const calculator = tool(
  ({ a, b, operation }) => {
    switch (operation) {
      case 'add':
        return `${a + b}`
      case 'subtract':
        return `${a - b}`
      case 'multiply':
        return `${a * b}`
      case 'divide':
        return b !== 0 ? `${a / b}` : 'Error: division by zero'
    }
  },
  {
    name: 'calculator',
    description: 'Perform basic arithmetic operations (add, subtract, multiply, divide)',
    schema: z.object({
      a: z.number().describe('The first number'),
      b: z.number().describe('The second number'),
      operation: z
        .enum(['add', 'subtract', 'multiply', 'divide'])
        .describe('The arithmetic operation to perform'),
    }),
  },
)

File: demos/crewai-agent/src/tools/calculator.py
from crewai.tools import tool


@tool("calculator")
def calculator(a: float, b: float, operation: str) -> str:
    """Perform basic arithmetic operations (add, subtract, multiply, divide).

    Args:
        a: The first number.
        b: The second number.
        operation: The operation to perform — one of: add, subtract, multiply, divide.
    """
    match operation:
        case "add":
            return str(a + b)
        case "subtract":
            return str(a - b)
        case "multiply":
            return str(a * b)
        case "divide":
            return str(a / b) if b != 0 else "Error: division by zero"
        case _:
            return f"Unknown operation: {operation}"
